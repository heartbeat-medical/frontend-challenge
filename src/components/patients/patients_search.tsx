import { FunctionComponent, useState, useEffect } from "react";
import { Patient, PatientSearchQuery } from "./patients";
import { Loading } from "../loading/loading";
import useToastContext from "../../hooks/useToastContext";

type props = {
  loadPatients: (query: PatientSearchQuery) => Promise<Patient[]>;
  onResults: (ps: Patient[]) => void;
};

export const PatientsSearch: FunctionComponent<props> = ({
  loadPatients,
  onResults,
}) => {
  const [query, updateQuery] = useState("");
  const [isLoading, setIsLoading] = useState(false);
  const { addToast } = useToastContext()

  useEffect(() => {
    const makeRequest = () => {
      setIsLoading(true);
      const sq: PatientSearchQuery = {
        name: query,
        ehrID: query,
        id: query,
      };
      
      loadPatients(sq)
        .then((ps) => {
          onResults(ps);
        })
        .catch((err) => {
          addToast({
            title: "Error Loading Patients",
            message: err,
            status: "error"
          });
        })
        .finally(() => {
          setIsLoading(false);
        });
    };

    makeRequest();
  }, [query, loadPatients, onResults, addToast])


  const handleChange = (e: any) => {
    updateQuery(e.target.value);
  }

  return (
    <div>
      {isLoading && <Loading />}
      <input
        onChange={handleChange}
      />
    </div>
  );
};
