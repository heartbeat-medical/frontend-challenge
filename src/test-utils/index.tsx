import React from "react";
import { render, RenderOptions } from "@testing-library/react";

interface CustomRenderOptions extends RenderOptions {
  initialHistory?: string[];
}

const customRender = (ui: any, options: CustomRenderOptions = {}) => {
  const { initialHistory, ...rest } = options;

  // Provide any context that the components may be expecting
  const Wrapper: React.FC = ({ children }) => <>{children}</>;
  return render(ui, { wrapper: Wrapper, ...rest });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
