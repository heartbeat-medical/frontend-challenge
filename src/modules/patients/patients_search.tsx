import { FunctionComponent as FC, useState } from "react";
import { Patient, PatientSearchQuery } from "./patients";
import { ToastBox, Loader } from "../../components";
import "./patients.css";

type props = {
  loadPatients: (query: PatientSearchQuery) => Promise<Patient[]>;
  onResults: (ps: Patient[]) => void;
};

const PatientsSearch: FC<props> = ({ loadPatients, onResults }) => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState("");
  const makeRequest = (inputValue: string) => {
    setIsLoading(true);
    const sq: PatientSearchQuery = {
      name: inputValue,
      ehrID: inputValue,
      id: inputValue,
    };
    loadPatients(sq)
      .then((ps) => {
        setIsLoading(false);
        setError("");
        onResults(ps);
      })
      .catch((err) => {
        setIsLoading(false);
        setError(err);
      });
  };
  return (
    <>
      <form>
        <label htmlFor="search-user">
          <input
            name="search-user"
            placeholder="Search by name, emp ID"
            onChange={(e) => makeRequest(e.target.value)}
          />
        </label>
      </form>
      <Loader isLoading={isLoading} />
      {error && <ToastBox title={error && error.toString()} status="error" />}
    </>
  );
};

export default PatientsSearch;
