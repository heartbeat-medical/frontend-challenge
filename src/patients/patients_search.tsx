import { FunctionComponent, useState } from "react";
import { Patient, PatientSearchQuery } from "./patients";

type props = {
  loadPatients: (query: PatientSearchQuery) => Promise<Patient[]>;
  onResults: (ps: Patient[]) => void;
};

export const PatientsSearch: FunctionComponent<props> = ({
  loadPatients,
  onResults,
}) => {
  const [query, updateQuery] = useState("");
  const makeRequest = () => {
    const sq: PatientSearchQuery = {
      name: query,
      ehrID: query,
      id: query,
    };
    loadPatients(sq)
      .then((ps) => onResults(ps))
      .catch((err) => alert(err));
  };
  return (
    <div>
      <label htmlFor="search-user">
        <input
          name="search-user"
          placeholder="Search by name, emp ID"
          onChange={(e) => {
            updateQuery(e.target.value);
            makeRequest();
          }}
        />
      </label>
    </div>
  );
};

type psearchboxprops = {
  onQueryChange: (query: PatientSearchQuery) => void;
};
