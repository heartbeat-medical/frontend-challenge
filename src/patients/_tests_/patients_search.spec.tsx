import { fireEvent, render, waitFor, cleanup } from "../../test-utils";
import { PatientsSearch } from "../patients_search";

describe("Patient Search test case", () => {
  let loadPatientMock: jest.Mock<any, any>;
  let onResultsMock: jest.Mock<any, any>;
  beforeEach(() => {
    loadPatientMock = jest.fn().mockImplementation(() => Promise.resolve());
    onResultsMock = jest.fn();
  });
  afterEach(() => {
    cleanup();
  });
  it("matches snapshot", () => {
    const { asFragment } = render(<PatientsSearch loadPatients={loadPatientMock} onResults={onResultsMock} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
