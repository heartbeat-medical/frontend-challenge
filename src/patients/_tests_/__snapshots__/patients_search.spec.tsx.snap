// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Patient Search test case matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      for="search-user"
    >
      <input
        name="search-user"
        placeholder="Search by name, emp ID"
      />
    </label>
  </div>
</DocumentFragment>
`;
